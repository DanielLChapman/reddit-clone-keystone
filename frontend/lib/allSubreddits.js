export const frontpage = [
    'Tech',
    'Videos',
    'Funny',
    'Politics',
    'Sports',
    'Entertainment',
    'News'
]