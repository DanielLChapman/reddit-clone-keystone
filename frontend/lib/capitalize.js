export default function capitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
  }
  

export function uniqBy(a, key) {
    var seen = {};
    return a.filter(function(item) {
        var k = JSON.stringify(item);
        return seen.hasOwnProperty(k) ? false : (seen[k] = true);
    })
}