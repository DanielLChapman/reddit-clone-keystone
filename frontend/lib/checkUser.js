export default function checkUserToSubreddit(user, moderators, owner) {
    let toTest = [owner, ...moderators];

    //check if user is the owner
    return toTest.some((x)=> {
        if (x.username.toLowerCase() === user.username.toLowerCase()) {
            return true;
        }
    })
}