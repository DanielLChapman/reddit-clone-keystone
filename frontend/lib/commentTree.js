let json = {};

class TreeNode {
    //constructors name and data for displaying information
    //parent incase we wanted to backwards traverse
    constructor(content,parent, id, comment) {
        this.content = content;
        this.parent = parent;
        this.descendents = [];
        this.id = id;
        this.commentObject = comment
    }

    setContent(value) {
        this.content = value;
    }
    setParent(value) {
        this.parent = value;
    }

    setID(value) {
        this.id = value;
    }
}


function makeid(length) {
    var result           = '';
    var characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for ( var i = 0; i < length; i++ ) {
      result += characters.charAt(Math.floor(Math.random() * 
 charactersLength));
   }
   return result;
}


export default class Tree {
    constructor() {
        this.root = new TreeNode('Root', null, 1, null);
        for (let x = 0; x < 10; x++) {
            let y = new TreeNode(makeid(10), 1, Math.floor(Math.random() * 10000), {});
            this.root.descendents.push(y);
        }
        let y = new TreeNode(makeid(10), 1, 55, {});

        let py = new TreeNode(makeid(10), 55, 550, {});
        y.descendents.push(py);
        this.root.descendents.push(y);
    }

    addNew = function(content, parentId, commentId, comment) {
        //find parent
        let parent = this.contains(parentId, this.traverseBF);

        let y = new TreeNode(content, parentId, commentId, comment);

        parent.descendents.push(y); 
    }


    contains = function(id, traversal) {
        let x = traversal(this, id);
        return x;
    }

    //breadth-first search with an array instead of a queue
    traverseBF = function(tree, id) {
        var queue = [];
        let descendents = tree.root.descendents
        queue = [...descendents];
        let currentNode = queue.shift();

        while (currentNode) {
            
            for (var i = 0; i < currentNode.descendents.length; i++) {
                queue.push(currentNode.descendents[i]);
            }	

            if (currentNode.id === id) {
                
                return currentNode;
            }
            currentNode = queue.shift();
        }
    }
		
    
    newPost(id, content, comment, parentId) {
        let newNode = new TreeNode(content, parentId, id, comment);
        //find parent
        let parent;

        if (parent === 1) {
            parent = root;
        }

        //start with root, go through descendents

        //if the parent id is in allBelowIds

        //loop through all descedents until parent is found

        //add to descendents

        //sort all descedents it was added to based on comment.total votes
    }



}