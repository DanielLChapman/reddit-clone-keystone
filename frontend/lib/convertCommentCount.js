export default convertCommentCount = (comments) => {
    if (comments < 1000) {
        return comments;
    }
    if (comments < 1000000) {
        let newVal = comments / 1000;
        newVal = newVal.toFixed(1);
        return newVal + "k"
    }
    else {
        let newVal = comments / 1000000;
        newVal = newVal.toFixed(1);
        return newVal + "milion"
    }
}