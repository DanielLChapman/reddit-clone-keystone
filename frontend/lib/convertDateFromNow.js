export default function convertDateFromNow(date) {
    let today = new Date();
    let calculatedSeconds = Math.abs((today - new Date(date))/1000);

    if (calculatedSeconds >= 31556952) {
        return `${Math.floor(calculatedSeconds/31556952)} year${Math.floor(calculatedSeconds/31556952) >= 2 ? 's': ''}  ago`;
    } else if (calculatedSeconds >= 2592000) {
        return `${Math.floor(calculatedSeconds/2592000)} month${Math.floor(calculatedSeconds/2592000) >= 2 ? 's': ''} ago`;
    } else if (calculatedSeconds >= 86400) {
        return `${Math.floor(calculatedSeconds/86400)} day${Math.floor(calculatedSeconds/86400) >= 2 ? 's': ''} ago`
    } else if (calculatedSeconds >= 3600) {
        return `${Math.floor(calculatedSeconds/3600)} hour${Math.floor(calculatedSeconds/3600) >= 2 ? 's': ''} ago`
    } 
    return `${Math.floor(calculatedSeconds/60)} minute${Math.floor(calculatedSeconds/60) >= 2 ? 's': ''} ago`
}