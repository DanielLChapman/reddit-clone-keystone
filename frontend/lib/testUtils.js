import casual from 'casual';

  
// seed it so we get consistent results
casual.seed(777);

const fakeUser = (overrides) => (
    {
    __typename: 'User',
    id: '4234',
    name: casual.name,
    username: casual.name,
    owner: [],
    email: casual.email,
    permissions: ['ADMIN'],
    moderating: [],
    subreddits: [],
    posts: [],
    postvotes: [],
    firstSignin: false,
    comments: [],
    commentvotes:[],
    ...overrides,
});

const fakeSubreddit = (overrides) => (
    {
        __typename: 'Subreddit',
        id: casual.uuid,
        name: casual.string,
        title: casual.string,
        slug: casual.string,
        sidebar: casual.string,
        rules: casual.string,
        description: casual.string,
        status: 'PUBLIC',
        ownder: fakeUser(),
        moderators: [],
        subscriber: [],
        posts: [],
        createdAt: Date.now(),
    }
);

const fakeLinkPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'www.google.com',
    type: 'Link',
    comments: [],
    removed: 'False',
    id: casual.uuid,


});

const fakeMediaPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'https://cdn.fileinfo.com/img/ss/sm/png_79.jpg',
    type: 'Link',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeTextPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    content: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: '',
    type: 'Text',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeComment = (overrides) => (
    {
    __typename: 'Comment',
    content: casual.sentence,
    user: fakeUser(),
    post: fakeTextPost(),
    createdAt: Date.now(),
    votes: [],
    parent: null,
    id: casual.uuid,
    children: [],

});


// Fake LocalStorage
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  }



export {

    LocalStorageMock,
    fakeLinkPost,
    fakeComment,
    fakeMediaPost,
    fakeSubreddit,
    fakeTextPost,
    fakeUser,

  };