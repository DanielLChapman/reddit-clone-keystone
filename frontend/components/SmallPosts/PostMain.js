import React from 'react';
import PropTypes from 'prop-types';
import { FaLongArrowAltUp, FaLongArrowAltDown } from 'react-icons/fa';
import Link from 'next/link';
import {convertDateFromNow} from '../../lib/convertDateFromNow';

function PostMain(props) {
    console.log(props);
    return (
        <section className="reddit-post" onClick={() => {window.location.href=`/r/`}}>
        <section className="reddit-post-left">
         {/*upvote, downvote and current score*/}
         <FaLongArrowAltUp />
         <br />
         <span></span>
         <br />
         <FaLongArrowAltDown />
        </section>
        <section className="reddit-post-right">
            <section className="reddit-post-right-top">
                <Link href={`../r/`}>
                    <a><b>r/kjnjknnk</b></a>
                </Link>
                <span>&nbsp;â€¢&nbsp;</span>
                <span>Posted by&nbsp;
                    <Link href={`../u/`}>
                    <a>u/sadsdasdasdas</a>
                </Link>
                </span>
                &nbsp;
                <span>Posted </span>
            </section>
            <section className="reddit-post-right-bottom">
                
            </section>
            <section className="reddit-post-right-bottom-footer">
  
            </section>

        </section>

    </section>
    );
}

PostMain.propType = {
    post: PropTypes.object.isRequired
}

export default PostMain;