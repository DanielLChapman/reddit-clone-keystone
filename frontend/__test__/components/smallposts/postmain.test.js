import { render, screen } from '@testing-library/react';
import PostMain from '../../../components/SmallPosts/PostMain';
import { fakeLinkPost, fakeMediaPost, fakeTextPost, fakeUser } from '../../../lib/testUtils';
import {MockedProvider} from '@apollo/react-testing';

const post = fakeTextPost();
const postMedia = fakeLinkPost();
const postMediaImage = fakeMediaPost();
const user = fakeUser();

describe('<PostMain />', () => {
    it('renders out the title and user', () => {
        const { container, debug } = render(
        <MockedProvider>
            <PostMain post={post} user={user} />
        </MockedProvider>);
        

        expect(screen.getByText('Posted by')).toBeInTheDocument();
        expect(screen.getByText(`Posted 0 minutes ago`)).toBeInTheDocument();
        expect(screen.getByText(`r/${post.subreddit.name}`)).toBeInTheDocument();
        expect(screen.getByText(`u/${post.user.username}`)).toBeInTheDocument();
        expect(screen.getByText(`${post.title}`)).toBeInTheDocument();
        
        let comments = container.querySelector('.reddit-post-right-bottom-footer .subreddit-link');

        expect(comments.textContent).toBe("0 Comments");

        const link = container.querySelectorAll('a');

        //link to user
        //link to subreddit
        //link to comments
        expect(link.length).toBe(3);
    });

    it ('renders out an up and downvote arrow', () => {
        const { container, debug } = render(
        <MockedProvider>
            <PostMain post={post} user={user} />
        </MockedProvider>);

        const leftSide = container.querySelector('.reddit-post .reddit-post-left');
        expect(leftSide).toBeDefined();
        let svg = leftSide.querySelectorAll('svg');

        expect(svg.length).toBe(2);

        expect(leftSide.textContent).toBe('0');
    });

    it('correctly renders the right post, media vs text', () => {
        let { container, debug } = render(
            <MockedProvider>
                <PostMain post={post} user={user} />
            </MockedProvider>);

        expect(container.querySelectorAll('.reddit-post-right-bottom-description').length).toBe(1);

        let { container: container2, debug: debug2 } = render(
            <MockedProvider>
                <PostMain post={postMedia} user={user} />
            </MockedProvider>);

        expect(container.querySelectorAll('.media-output-small-post').length).toBe(0);
        expect(container2.querySelectorAll('.media-output-small-post').length).toBe(1);
        expect(container2.querySelectorAll('.media-output-small-post a').length).toBe(1);
    })

    it('renders and matches the snapshop', () => {
        let { container, debug } = render(
            <MockedProvider>
                <PostMain post={post} user={user} />
            </MockedProvider>);

        expect(container).toMatchSnapshot();
    });

    it('renders the image properly', () => {
        let { container, debug } = render(
            <MockedProvider>
                <PostMain post={postMediaImage} user={user} />
            </MockedProvider>);

            const img = screen.getByAltText(postMediaImage.title);
            expect(img).toBeInTheDocument();
    })
})