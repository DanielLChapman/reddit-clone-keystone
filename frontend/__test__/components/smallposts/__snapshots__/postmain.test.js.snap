// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostMain /> renders and matches the snapshop 1`] = `
<div>
  <section
    class="reddit-post"
    style="display: block;"
  >
    <section
      class="reddit-post-left"
    >
      <svg
        class="upvote-arrow "
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 256 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M88 166.059V468c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12V166.059h46.059c21.382 0 32.09-25.851 16.971-40.971l-86.059-86.059c-9.373-9.373-24.569-9.373-33.941 0l-86.059 86.059c-15.119 15.119-4.411 40.971 16.971 40.971H88z"
        />
      </svg>
      <div
        aria-disabled="false"
        class="upvote-arrow-box"
      />
      <br />
      <span>
        0
      </span>
      <br />
      <svg
        class="upvote-arrow "
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 256 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M168 345.941V44c0-6.627-5.373-12-12-12h-56c-6.627 0-12 5.373-12 12v301.941H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.569 9.373 33.941 0l86.059-86.059c15.119-15.119 4.411-40.971-16.971-40.971H168z"
        />
      </svg>
      <div
        aria-disabled="false"
        class="downvote-arrow-box"
      />
    </section>
    <section
      class="reddit-post-right"
    >
      <section
        class="reddit-post-right-top"
      >
        <a
          href="/r/illum saepe ut assumenda sint quia sed"
        >
          <b
            class="subreddit-link"
          >
            r/
            ipsa tempora aut sint aperiam occaecati illum
          </b>
        </a>
        <span>
           • 
        </span>
        <span>
          Posted by 
          <a
            class="subreddit-link"
            href="/user/Mrs. Gideon Block"
          >
            u/
            Mrs. Gideon Block
          </a>
        </span>
         
        <span>
          Posted 
          0 minutes ago
        </span>
      </section>
      <section
        class="reddit-post-right-bottom"
      >
        <h4>
          Facilis corporis sed veritatis.
        </h4>
        <section
          class="reddit-post-right-bottom-description "
        >
          <span>
            <p>
              Praesentium magnam fugiat error molestiae laborum est.
            </p>
          </span>
        </section>
      </section>
      <section
        class="reddit-post-right-bottom-footer"
      >
        <a
          class="subreddit-link"
          href="/r/illum saepe ut assumenda sint quia sed/comments/7760116b-dd26-46da-a05a-a010e05cef40/explicabo dolores ipsam laboriosam non itaque voluptas"
        >
          0
           Comments
        </a>
        
      </section>
    </section>
  </section>
</div>
`;
